modules/scope3/rtd/module.go

```diff
@@ -313,50 +300,108 @@ func (m *Module) HandleAuctionResponseHook(
 	// Add segments to the auction response
 	ret.ChangeSet.AddMutation(
 		func(payload hookstage.AuctionResponsePayload) (hookstage.AuctionResponsePayload, error) {
-			// Add Scope3 segments to the response ext so publisher can use them
-			if payload.BidResponse.Ext == nil {
-				payload.BidResponse.Ext = json.RawMessage("{}")
-			}
-
-			var extMap map[string]interface{}
-			if err = jsonutil.Unmarshal(payload.BidResponse.Ext, &extMap); err != nil {
-				extMap = make(map[string]interface{})
-			}
-
 			// Add segments as individual targeting keys for GAM integration
 			if m.cfg.AddToTargeting {
-				prebidMap, ok := extMap["prebid"].(map[string]interface{})
-				if !ok {
-					prebidMap = make(map[string]interface{})
-					extMap["prebid"] = prebidMap
-				}
-				targetingMap, ok := prebidMap["targeting"].(map[string]interface{})
-				if !ok {
-					targetingMap = make(map[string]interface{})
-					prebidMap["targeting"] = targetingMap
-				}
 				// Add each segment as individual targeting key
 				for _, segment := range segments {
 					if strings.HasPrefix(segment, scope3MacroKeyPlusSeparator) {
 						macroKeyVal := strings.Split(segment, scope3MacroSeparator)
 						if len(macroKeyVal) != 2 {
 							continue
 						}
-						targetingMap[macroKeyVal[0]] = macroKeyVal[1]
+						newPayload, err := sjson.SetBytes(payload.BidResponse.Ext, "prebid.targeting."+macroKeyVal[0], macroKeyVal[1])
+						if err != nil {
+							return payload, err
+						}
+						payload.BidResponse.Ext = newPayload
 					} else {
-						targetingMap[segment] = "true"
+						if m.cfg.SingleSegmentKey != "" {
+							currentVal := gjson.GetBytes(payload.BidResponse.Ext, "prebid.targeting."+m.cfg.SingleSegmentKey)
+							if currentVal.Exists() {
+								newPayload, err := sjson.SetBytes(payload.BidResponse.Ext, "prebid.targeting."+m.cfg.SingleSegmentKey, currentVal.String()+","+segment)
+								if err != nil {
+									return payload, err
+								}
+								payload.BidResponse.Ext = newPayload
+							} else {
+								newPayload, err := sjson.SetBytes(payload.BidResponse.Ext, "prebid.targeting."+m.cfg.SingleSegmentKey, segment)
+								if err != nil {
+									return payload, err
+								}
+								payload.BidResponse.Ext = newPayload
+							}
+						} else {
+							newPayload, err := sjson.SetBytes(payload.BidResponse.Ext, "prebid.targeting."+segment, "true")
+							if err != nil {
+								return payload, err
+							}
+							payload.BidResponse.Ext = newPayload
+						}
```

Comment:
I feel like this could be simplified significantly by having the logic choose or override the key and value and doing the newpayload and error checking only once.